### DISCUSSION TITLE
rejectOnNotFound throws Error when non-required relation is null

### DISCUSSION THREAD

**Discussion by rtnolan** [prisma/prisma#8900] [Q&A] [answered]
Hi guys,

I'm attempting to throw an Error when `findUnique()` doesn't return a value for a given model. Right now my model looks like this:
```prisma
model User {
  id          String     @id @default(cuid())
  email       String     @unique
  infoId      String     @unique
  Info        Info?      @relation(fields: [infoId], references: [id])
}

model Info {
  id          String @id @default(cuid())
  User        User?
}
```
I've tried to set my client up in the following ways:
```javascript
export const prisma = new PrismaClient({
    rejectOnNotFound: {
        findUnique: true
    }
});

// and

export const prisma = new PrismaClient({
    rejectOnNotFound: {
        findUnique: {
            User: (err) => new Error("Could not find user.")
        }
    }
});
```
`Info` isn't a required relation but when it returns null it's tripping an Error on our gql client (`urql`) but also returning the User as expected. Output:
```json
{
  "errors": [
    {
      "message": "Could not find user.",
      "locations": [{ "line": 12, "column": 5 }],
      "path": ["getUser", "Info"],
      "extensions": {
        "code": "INTERNAL_SERVER_ERROR",
        "exception": {
          "clientVersion": "2.29.0",
          "stacktrace": [
            "Error: Could not find user.",
            "..."
          ]
        }
      }
    }
  ],
  "data": {
    "getUser": {
      "id": "ckfqy3w2q022813mawgzib8x4",
      "email": "test123@gmail.com",
      "Info": null,
      "__typename": "User"
    }
  }
}

```
Is there a way to configure the Prisma Client to throw an Error if it doesn't find the resource it's looking for but ignore relations being null? Any help is appreciated, thank you.

**Comment by ryands17** [comment, marked as answer]
@tnolan8 👋 

Your Prisma schema is incorrect. You cannot have a foreign key that is required but the virtual field that is null. In the above schema, `infoId` is required but `Info` is not and that's incorrect.

You need to change your schema to this (notice the `?` on `infoId`):

```prisma
model User {
  id     String @id @default(cuid())
  email  String @unique
  infoId String? @unique
  Info   Info?  @relation(fields: [infoId], references: [id])
}

model Info {
  id   String @id @default(cuid())
  User User?
}
```

And it will work fine.

**Reply by rtnolan** [reply]
Hi @ryands17,

That was actually a copy and paste error on my behalf, `infoId` is indeed optional on my schema, apologies.

**Reply by ryands17** [reply]
Works fine for me, are you sure it's a Prisma issue and not one in your GraphQL resolver? I tried it with fetching a user and using `include` for fetching `info`. The user that is found doesn't throw an error if the relation is null and a non-existing user successfully throws an error if fetched.

**Reply by rtnolan** [reply]
The resolver is pretty straightforward, so I don't think it's a problem with that. When I'm executing the query through Apollo's Sandbox it looks like it returns fine with no `Errors` object but using `urql` for some reason it's returning with an `Errors` object also. I'll see if I can mimic the problem in a sample repo but I don't think it's on Prisma's side.

**Reply by rtnolan** [reply]
https://github.com/tnolan8/urql-prisma - If you're at all interested, I'm not totally sure where the problem is. I'll see if I can find anything out from the `urql`/`typegraphql` communities.

**Reply by ryands17** [reply]
Could you try it without TypeGraphQL all together? i.e. creating a simple seed file and then running the operations there. If it works, that means it's a TypeGraphQL issue.

**Reply by rtnolan** [reply]
It does indeed look like it's an issue with TypeGraphQL, I'll open an issue there, thanks for your help @ryands17  👋

**Comment by rtnolan** [comment]
Hey @ryands17 👋  

I just had a quick discussion over on the TypeGraphQL repo and I'm no longer sure the problem lies there. TypeGraphQL generates Field Resolvers similar to this:
```typescript
@TypeGraphQL.Resolver((_of) => User)
export class UserRelationsResolver {
  @TypeGraphQL.FieldResolver((_type) => Info, {
    nullable: true,
  })
  async Info(
    @TypeGraphQL.Root() user: User,
    @TypeGraphQL.Ctx() ctx: any
  ): Promise<Info | null> {
    return getPrismaFromContext(ctx)
      .user.findUnique({
        where: {
          id: user.id,
        },
      })
      .Info({});
  }
}
```
The problem seems to be the fact that it the relation `.Info({})` resolves to `null` and the Prisma Client is set up to throw and Error if `user.findUnqiue({...})` returns a `null` value. This seems like a bug on the Prisma side right?

**Reply by ryands17** [reply]
Why would that be an error? `rejectOnNotFound` will throw an error if `user.findUnique` returns `null` as that's what you specified in `PrismaClient`. So this is completely normal behaviour as per the configuration.

**Reply by rtnolan** [reply]
Okay I guess that it's not possible then to use `rejectOnNotFound` in this case then! Thank you.

**Reply by ryands17** [reply]
`rejectOnNotFound` is for the main model and not its relations. If the user doesn't exist, it will throw regardless of the relation being there or not 🙂

**Reply by rtnolan** [reply]
If `rejectOnNotFound` is for the main model then I would have thought this should work. The actual call to `prisma.user.findUnique({...})` returns a User it's just the fact that the Field Resolver for `Info` uses `prisma.user.findUnique({..}).Info({})` which resolves too `null`. I would have thought that this would have been fine as the `findUnqiue({...})` call does return something it's just `.Info({})` that is `null`. I may have just misunderstood usage.

**Reply by ryands17** [reply]
Got your point. This seems like an issue when you use the [Fluent API](https://www.prisma.io/docs/concepts/components/prisma-client/relation-queries#fluent-api). It would be great if you could create one [here](https://github.com/prisma/prisma/issues/new?assignees=&labels=kind%2Fbug&template=bug_report.yml) with the above schema and reproduction step.

As a workaround, you can use `include` like this to fetch the relation and it will not throw an error if the user exists:

```ts
let user = await prisma.user.findUnique({
    where: { id: u.id },
    include: { Info: true },
  })

console.log(user.Info)
```

**Reply by rtnolan** [reply]
Will open an issue now, cheers for the help @ryands17.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
